core/src/main/scala/org/apache/spark/sql/streaming/SnappySinkCallback.scala

```diff
@@ -218,11 +217,20 @@ class DefaultSnappySinkCallback extends SnappySinkCallback {
         index += 1
         contains
       }
-      val conflatedDf = if (otherCols.isEmpty) df.distinct()
-      else {
-        val exprs = otherCols.map(c => last(c).alias(c))
+
+      val conflatedDf: DataFrame = {
+        val exprs = otherCols.map(c => last(c).alias(c)) ++
+            Seq(count(lit(1)).alias(EVENT_COUNT_COLUMN))
+
+        // if event type of the last event for a key is insert and there are more than one
+        // events for the same key, then convert inserts to put into
+        val columns = df.columns.filter(_ != EVENT_TYPE_COLUMN).map(col) ++
+            Seq(when(col(EVENT_TYPE_COLUMN) === EventType.INSERT && col(EVENT_COUNT_COLUMN) > 1,
+              EventType.UPDATE).otherwise(col(EVENT_TYPE_COLUMN)).alias(EVENT_TYPE_COLUMN))
+
         df.groupBy(keyCols.head, keyCols.tail: _*)
             .agg(exprs.head, exprs.tail: _*)
+            .select(columns: _*)
             .select(df.columns.head, df.columns.tail: _*)
```

Comment:
Is the second select() required given that all required columns have already been projected in first?